src/snowflake/snowpark/functions.py

```diff
@@ -8700,3 +8711,135 @@ def make_interval(
+        wrap_row: A flag to determine if the current row must be wrapped in an instance of the :class:`Row` class.
+        kwargs: Additional arguments to register the UDTF used to apply the Python function.
+
+    Example 1::
+
+        >>> from snowflake.snowpark.types import IntegerType
+        >>> from snowflake.snowpark.functions import map
+        >>> df = session.create_dataframe([[10, "a"]], schema=["col1", "col2"])
+        >>> new_df = map(df, lambda row: row[0] * row[0], output_types=[IntegerType()])
+        >>> new_df.show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        ---------
+        <BLANKLINE>
+
+    Example 2::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()])
+        >>> new_df.show()
+        -----------------
+        |"C_1"  |"C_2"  |
+        -----------------
+        |a      |30     |
+        -----------------
+        <BLANKLINE>
+
+    Example 3::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()], output_column_names=['col1', 'col2'])
+        >>> new_df.show()
+        -------------------
+        |"COL1"  |"COL2"  |
+        -------------------
+        |a       |30      |
+        -------------------
+        <BLANKLINE>
+
+    Example 4::
+        >>> df = session.create_dataframe([10, 20, 30])
+        >>> new_df = map(df, lambda x: x * x, output_types=[IntegerType()], wrap_row=False)
+        >>> new_df.sort(new_df[0]).show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        |400    |
+        |900    |
+        ---------
+        <BLANKLINE>
+
+    Note:
+        The result of the `func` function must be either a scalar value or a tuple containing the same number of elements
+        as specified in the `output_types` argument.
+    """
+    if "packages" not in kwargs:
+        kwargs["packages"] = ["snowflake-snowpark-python"]
+    if len(output_types) == 0:
+        raise ValueError("output_types cannot be empty.")
+    input_types = [field.datatype for field in dataframe.schema.fields]
+    num_fields = len(dataframe.schema.fields)
+    output_schema = [
+        StructField(f"c_{i + 1}", output_types[i]) for i in range(len(output_types))
+    ]
+    output_cols = [
+        col(f"${i + 1}") for i in range(num_fields, num_fields + len(output_types))
+    ]
+    input_row_obj = snowflake.snowpark.Row(*dataframe.columns) if wrap_row else None
+
+    def wrap_result(result):
+        if isinstance(result, list) or isinstance(result, snowflake.snowpark.Row):
+            return tuple(result)
+        elif isinstance(result, tuple):
+            return result
+        else:
+            return (result,)
+
+    class _MapFunc:
+        def process(self, *argv):
+            if wrap_row:
+                argv = [input_row_obj(*argv)]
```

Comment:
```suggestion
            if wrap_row:
                argv = [Row(*argv)]
```

Reply:
@sfc-gh-aalam the goal of this code is to support scenarios where the user uses column names in the function that is applied to each row. For example:

```python
another_df = session.create_dataframe([('a',10), ('b', 20), ('d', 30)], 
                                      schema=StructType([StructField('NAME', StringType()),
                                                         StructField('NUMVAL', IntegerType())]))

r1 = map(another_df, lambda row: (row.NAME + "_", row.NUMVAL * row.NUMVAL), [StringType(), IntegerType()])
r1.show()
```

This is the reason for creating a `Row` object from the column names of the current dataframe (next comment).

If we do not keep this information we get the following error:

```
...
    raise AttributeError(f"Row object has no attribute {item}")
AttributeError: Row object has no attribute NAME
```

Reply:
I don't understand why we need to create `input_row_obj`. Maybe I don't understand `Row` correctly. There are two cases here.

Case 1: `wrap_row = False`
- `input_row_obj = None`
- We don't execute line 8822-8823:
```
if wrap_row:
    argv = [input_row_obj(*argv)]
```

Case 2: `wrap_row = True`
- `input_row_obj = Row`
- we execute line 8822-8823:
```
if wrap_row:
    argv = [Row(*argv)]
```

If that is the case, why do we even need to create `input_row_obj`. can't we just do
```
if wrap_row:
   argv = [Row(*argv)]
```

Reply:
We need `input_row_obj` to create Row object instances that have the same column names as the ones in the current DataFrame.  We want to support these scenarios:

```python
map(df, lambda row: (row.C1 + 1, row.C1 * 2), [IntegerType(), IntegerType()])
```

Notice that in `functions.py` in line 8810 we create `input_row_obj` like this:

```
input_row_obj = snowflake.snowpark.Row(*dataframe.columns) if wrap_row else None
```

As described in the second example for the [documentation](https://docs.snowflake.com/developer-guide/snowpark/reference/python/latest/snowpark/api/snowflake.snowpark.Row) ,   `Row` instances are ['callable'](https://github.com/snowflakedb/snowpark-python/blob/cef750a346d60ecb5b0fe4d3e7fe6e5ad2571a5e/src/snowflake/snowpark/row.py#L199) and can be used to create other `Row` instances.

Reply:
Okay. this makes sense to me. Thanks for clarifying.